### ISSUE TITLE
Ability to do a strict checkValid()?

### ISSUE THREAD

**Issue by ThomasWu** [lightbend/config#539]
I am trying to use checkValid() to validate user provided configuration vs our reference. I want to detect invalid config as early as possible, but I don't want to have a long and tedious validation logic for each config field since we already have a well defined reference config, which contains everything. However with the current implementation of checkValid(), the validation is basically failing if we want to have value type as string in the config file. I wonder if it is possible to have a strict validation that also detects the difference between string and all other types?

**Comment by havocp** [comment]
Can you give a more specific example of a situation, what's the value in the reference.conf and what value in the actual conf do you want to fail?

checkValid doesn't have a ton of information to work with I guess, since there's no full schema or anything

**Comment by ThomasWu** [comment]
The situation I am facing is very simple.
Assume what I have in reference is `a: "test"`. I want checkValid to fail if I receive a config like `a: 123` or `a: true`. 
Currently checkValid won't complain about this situation, but later if I am extracting values with getString() I will have a problem.

**Comment by havocp** [comment]
Yeah. It comes down to this function I guess: https://github.com/lightbend/config/blob/eeb95825bf3763959ddd073d20b11045d2c01131/config/src/main/java/com/typesafe/config/impl/SimpleConfig.java#L967-L999

And then number/boolean can be gotten as a string (I think getString actually works in that case?): https://github.com/lightbend/config/blob/eeb95825bf3763959ddd073d20b11045d2c01131/config/src/main/java/com/typesafe/config/impl/DefaultTransformer.java#L65-L68

There's some ambiguity not found in JSON because of unquoted strings; you can have a string with no quotes, but if that string happens to be `true` or `123` then we parse it as bool/number but convert it back to a string when asked... if the unquoted string is `abc` then it's parsed as a string to begin with.

If your code handles quoted `"123"` and calls getString, then I _think_ it would also handle unquoted `123` (which should get converted to a string, and then validated like `"123"` which is a string to start?)

**Comment by ThomasWu** [comment]
You're right, the default converter does work. I don't know why I thought it doesn't work. 
Another issue I have with current checkValid is that checkValid on `a: [1,2,3]` vs reference `a: "123"` actually passes, because the reference value type is string. If you flip these two, checkValid will complain the type difference.
I think this would be a bug then.

**Comment by havocp** [comment]
Hmm, yes. The code and the comment actually appear to be out of sync on that:
https://github.com/lightbend/config/blob/eeb95825bf3763959ddd073d20b11045d2c01131/config/src/main/java/com/typesafe/config/impl/SimpleConfig.java#L984-L988

i.e. "Assume a string can be gotten as any non-collection type" but there's no check here for collection types (lists, objects). I'm trying to think about whether it's a breaking change to fix this... it seems plausible to me that someone has an app that works now, that will break if we make this check more strict :-/ but it's probably ok at least in a major release bump.
